syntax = "proto3";

import "google/protobuf/any.proto";

import "mixing_state.proto";
import "cluster_state.proto";
import "hierarchy_prior.proto";
import "distribution.proto";

package bayesmix;


/*
 * This message represents the state of a Gibbs sampler for
 * a mixture model. All algorithms must be able to handle this
 * message, by filling it with the current state of the sampler
 * in the `get_state_as_proto` method.
 */
message AlgorithmState {

  // Represents the state of a single cluster of component of the mixture model.
  message ClusterState {
    oneof val {
      google.protobuf.Any custom_state = 2; // Custom Message defined by user
      UniLSState uni_ls_state = 3; // State of a univariate location-scale family
      MultiLSState multi_ls_state = 4; // State of a multivariate location-scale family
      LinRegUniLSState lin_reg_uni_ls_state = 5; // State of a linear regression univariate location-scale family
      FAState fa_state = 6; // State of a Mixture of Factor Analysers
    }
    int32 cardinality = 1; // How many observations are in this cluster
  }

  // Represents the current values of the Hyperparameters of the Hierarchy
  message HierarchyHypers {
    oneof val {
      google.protobuf.Any custom_state = 1;
      NIGDistribution nnig_state = 2;
      NWDistribution nnw_state = 3;
      MultiNormalIGDistribution lin_reg_uni_state = 4;
      NxIGDistribution nnxig_state = 5;
      FAPriorDistribution fa_state = 6;
      EmptyPrior empty_state = 7;
    }
  }

  repeated ClusterState cluster_states = 1; // The state of each cluster
  repeated int32 cluster_allocs = 2 [packed = true]; // Vector of allocations into clusters, one for each observation
  MixingState mixing_state = 3; // The state of the `Mixing`
  int32 iteration_num = 4; // The iteration number
  HierarchyHypers hierarchy_hypers = 5; // The current values of the hyperparameters of the hierarchy
}
